File: NPC/TSP/TSP.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <cmath>

using std::cout;
using std::cin;

class Algorithm{
public:
    double ans = INT8_MAX;
    std::string path;
    int ncities;
    std::vector<std::pair<double, double>> location;
    std::vector<std::vector<double>> C;
    std::vector<std::vector<double>> A;

    void read_file();
    void compute_len();
    void run_algo();
    int dp(int, int);

    Algorithm(std::string _p): path(_p){}
};

void Algorithm::read_file(){
    std::ifstream fh(path);
    fh >> ncities;
    location = std::vector<std::pair<double, double>>(ncities+1);
    C = std::vector<std::vector<double>>(ncities+1, std::vector<double>(ncities+1, 0));
    A = std::vector<std::vector<double>>((ncities+1), std::vector<double>(1<<(ncities+1), INT8_MAX));
    
    int idx=1;
    double x, y;
    while(fh >> x >> y){
        location[idx] = {x, y};
        idx++;
    }
    fh.close();
    cout << "done read.\n";
}

void Algorithm::compute_len(){
    for(int i=1; i<=ncities; i++){
        for(int j=1; j<=ncities; j++){
            std::pair<double, double> loc1 = location[i];
            std::pair<double, double> loc2 = location[j];
            C[i][j] = std::sqrt(std::pow(loc1.first-loc2.first, 2) + std::pow(loc1.second-loc2.second, 2));
            C[j][i] = C[i][j];
        }
    }

    cout << "done computing len.\n";
}

void Algorithm::run_algo(){
    A[1][1] = 0;

    for(int j=2; j<=ncities; j++){
        ans = std::min(ans, dp(j, (1<<(ncities+1))-1)+C[j][1]);
    }
}

int Algorithm::dp(int start, int vtx_comb){
    if(vtx_comb == (1<<start | 1)) return C[1][start];

    if(A[start][vtx_comb] != INT8_MAX) return A[start][vtx_comb];

    int res = INT8_MAX;
    for(int k=1; k<=ncities; k++){
        if(vtx_comb & (1<<k) && k!=start && k!=1){
            int new_state = vtx_comb & (~(1<<start));
            res = std::min(res, dp(k, new_state));
        }
    }
    A[start][vtx_comb] = res;
    return res;
}

int main(int argc, char* argv[]){
    Algorithm TSP(argv[1]);
    TSP.read_file();
    TSP.compute_len();
    TSP.run_algo();
    cout << TSP.ans << "\n";
}

File: NPC/TSP/TSP.java
import java.io.*;
import java.util.*;

public class TSP{
    static double MAX = 1e8;
    static double ans = MAX;
    static int ncities;
    static double[][] location;
    static double[][] C;
    static double[][] A;

    public static void compute_len(){
        for(int i=1; i<=ncities; i++){
            for(int j=1; j<=ncities; j++){
                double locix = location[i][1];
                double lociy = location[i][2];
                double locjx = location[j][1];
                double locjy = location[j][2];

                C[i][j] = Math.sqrt(Math.pow(locix-locjx, 2)+Math.pow(lociy-locjy, 2));
            }
        }
        System.out.println("done compute len.\n");
    }

    public static void run_algo(){
        for(int i=1; i<=ncities; i++){
            System.out.println("visting: " + i);
            ans = Math.min(ans, dp(i, (1<<(ncities+1))-1)+C[i][1]);
        }
        System.out.println("the optimal solution: "+ ans);
    }

    public static double dp(int start, int vtx_comb){
        // this method compute the min dist from source 1 to dest start visiting every vtx in vtx_comb once.
        if(vtx_comb == ((1<<start) | 1)) return C[1][start];
        if(A[start][vtx_comb]!=0) return A[start][vtx_comb];

        double res = MAX;

        for(int k=1; k<=ncities; k++){
            if((vtx_comb & (1<<k))!=0 && k!=start && k!=1){
                int new_state = vtx_comb& (~(1<<start));
                res = Math.min(res, dp(k, new_state)+C[k][start]);
            }
        }
        A[start][vtx_comb] = res;
        return res;
    }

    public static void main(String[] argv){
        try{
            Scanner sc = new Scanner(new File(System.console().readLine()));
            ncities = sc.nextInt();
            C = new double[ncities+1][ncities+1];
            A = new double[ncities+1][1<<(ncities+1)];
            location = new double[ncities+1][3];

            for(int i=1; i<=ncities; i++){
                for(int j=1; j<=2; j++){
                    location[i][j] = sc.nextDouble();
                }
            }

            compute_len();
            run_algo();
        } catch (FileNotFoundException e){
            System.out.println("File not found.\n");
        }
    }
}